server/src/main/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplate.java

```diff
@@ -352,10 +336,13 @@ public DataStreamTemplate(boolean hidden, boolean allowCustomRouting, IndexMode
             } else {
                 allowCustomRouting = false;
             }
-            if (in.getVersion().onOrAfter(Version.V_8_1_0)) {
-                indexMode = in.readOptionalEnum(IndexMode.class);
-            } else {
-                indexMode = null;
+            if (in.getVersion().onOrAfter(Version.V_8_1_0) && in.getVersion().before(Version.V_8_3_0)) {
+                // Accidentally included index_mode to binary node to node protocol in previous releases.
+                // (index_mode is removed and was part of code based when tsdb was behind a feature flag)
+                // (index_mode was behind a feature in the xcontent parser, so it could never actually used)
+                // (this used to be an optional enum, so just need to (de-)serialize a false boolean value here)
+                boolean value = in.readBoolean();
```

Comment:
I think this is the most important place to keep it enum.

Reply:
The index_mode is removed from the the composable index templates, the only reason that a boolean is read here is because index mode wasn't behind a feature flag in the binary serialisation code (it was in the xcontent serialization). 

The index mode is kept in the `DataSteam` class. 

So I think this is good?